fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/lifecycle/v1/ClientLifecycleEvents.java

```diff
@@ -0,0 +1,235 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.event.lifecycle.v1;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.world.ClientWorld;
+import net.minecraft.util.profiler.Profiler;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.api.event.lifecycle.v1.GameLifecycleCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.GameTickCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.block.entity.BlockEntityLoadCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.block.entity.BlockEntityUnloadCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.chunk.ChunkLoadCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.chunk.ChunkUnloadCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.entity.EntityLoadCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.entity.EntityUnloadCallback;
+import net.fabricmc.fabric.api.event.lifecycle.v1.world.WorldTickCallback;
+
+@Environment(EnvType.CLIENT)
+public final class ClientLifecycleEvents {
+	private ClientLifecycleEvents() {
+	}
+
+	/**
+	 * Called when a client ticks.
+	 */
+	public static final Event<GameTickCallback<MinecraftClient>> CLIENT_TICK = EventFactory.createArrayBacked(GameTickCallback.class, callbacks -> client -> {
+		if (EventFactory.isProfilingEnabled()) {
+			final Profiler profiler = client.getProfiler();
+			profiler.push("fabricServerTick");
+
+			for (GameTickCallback<MinecraftClient> event : callbacks) {
+				profiler.push(EventFactory.getHandlerName(event));
+				event.onTick(client);
+				profiler.pop();
+			}
+
+			profiler.pop();
+		} else {
+			for (GameTickCallback<MinecraftClient> event : callbacks) {
+				event.onTick(client);
+			}
+		}
+	});
+
+	/**
+	 * Called when a ClientWorld ticks.
+	 */
+	public static final Event<WorldTickCallback<ClientWorld>> WORLD_TICK = EventFactory.createArrayBacked(WorldTickCallback.class, callbacks -> world -> {
+		if (EventFactory.isProfilingEnabled()) {
+			final Profiler profiler = world.getProfiler();
+			profiler.push("fabricClientWorldTick");
+
+			for (WorldTickCallback<ClientWorld> callback : callbacks) {
+				profiler.push(EventFactory.getHandlerName(callback));
+				callback.onTick(world);
+				profiler.pop();
+			}
+
+			profiler.pop();
+		} else {
+			for (WorldTickCallback<ClientWorld> callback : callbacks) {
+				callback.onTick(world);
+			}
+		}
+	});
+
+	/**
+	 * Called when a chunk is loaded into a ClientWorld.
```

Comment:
For load/unload the timing is crucial and should be explained. Ideally all load events occur after the subject has been added to the world/chunk with the world+chunk being normally usable and obtainable via MinecraftServer. This ensures that e.g. an area search can be done without fragile postponing workarounds. Unload should happen while still reachable for similar reasons.

Additionally, access to the de-/serialization phases would be useful such that custom data can be embedded in the vanilla nbt tree as it gets read or written. It is useful for persistent data attachments to chunk or world, e.g. to store worldgen metadata for retrogen implementations.

Reply:
I've had a discussion on discord related to a `Save/LoadEntityDataCallback` and the concenus was it isn't very useful unless we have a way to attach data to an entity (via something like CCA) or mixin to said entity to access the data you wish to store.